Make subpixel-positioned iframes paint and hit-test correctly

smfr opened the pull request:
#### 79b2e213326cb5035fddf5ef403a0d50fe6b1dd5
<pre>
Make subpixel-positioned iframes paint and hit-test correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=294571">https://bugs.webkit.org/show_bug.cgi?id=294571</a>
<a href="https://rdar.apple.com/152114545">rdar://152114545</a>

Reviewed by Alan Baradlay.

IFrames are Widgets, and Widgets live in an integer-based hierarchy, so have historically
painted and hit-tested using integer coordinates. However, iframes are often positioned on
subpixel boundaries, which leads to pixel gaps. Some of these are very obvious, e.g. an
iframe inside a `position:fixed; bottom: 0` element that should align seamlessly with UI.

However, we can adjust iframes for painting and hit-testing to overcome the integral
Widget limitation. This PR does this in various places.

First, `RenderWidget::paintContents()` adjusts the CTM to account for a subpixel position,
which fixes painting gaps (tested by iframe-subpixel-painting.html).

Second, the compositing code needs to handle subpixel positioning of the iframe&apos;s layers,
fixed in `RenderLayerBacking::updateAfterWidgetResize()` and tested by
iframe-subpixel-compositing.html.

Finally, hit-testing needs to maintain fractional coordinates;
`documentPointForWindowPoint()` is fixed to use FloatPoint coordinate mapping, and
HitTestLocation no longer rounds the origin of m_boundingBox which is what&apos;s used by the
`nodeAtPoint()` code.

Removing the flooring of the point in the HitTestLocation constructor caused
`css/css-transforms/transform-scale-hittest.html` to fail; this hit-tests a pixel above an
element with a 100x scale. The existing code would map the 1x1px hit-test area through the
scale, resulting in a 100x100px hit-test area; fix to map through the inverse transform to
maintain the 1x1px hit-test area in screen space. This is a merge of
<a href="https://chromium.googlesource.com/chromium/src.git/+/c47a6f0b4a592c2383fb553b6169f89ab01d851d">https://chromium.googlesource.com/chromium/src.git/+/c47a6f0b4a592c2383fb553b6169f89ab01d851d</a>

* LayoutTests/TestExpectations: imported/w3c/web-platform-tests/css/selectors/focus-visible-018-2.html
fails; <a href="https://github.com/web-platform-tests/wpt/pull/53241">https://github.com/web-platform-tests/wpt/pull/53241</a> exists to fix this.
* LayoutTests/compositing/iframes/hidpi-iframe-subpixel-compositing-expected.html: Added.
* LayoutTests/compositing/iframes/hidpi-iframe-subpixel-compositing.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-hit-test-expected.txt: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-hit-test.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-painting-expected.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-painting.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/elementFromPoint-subpixel-expected.txt:
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2-expected.txt:
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2.html: Tweak the test; browsers
have different results.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::documentPointForWindowPoint):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/rendering/HitTestLocation.cpp:
(WebCore::HitTestLocation::HitTestLocation): No longer floor the point.
(WebCore::HitTestLocation::move): No longer convert to IntRect.
* Source/WebCore/rendering/HitTestingTransformState.cpp:
(WebCore::HitTestingTransformState::boundsOfMappedQuad const):
* Source/WebCore/rendering/HitTestingTransformState.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTestLayerByApplyingTransform):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterWidgetResize):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::frameViewDidChangeLocation):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::paintContents):

Canonical link: <a href="https://commits.webkit.org/296407@main">https://commits.webkit.org/296407@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9974033a603c7206123021c222219d0a405cd9a8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108387 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28048 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113596 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58808 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110350 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28737 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36598 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82304 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111335 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22791 "Found 1 new test failure: http/tests/security/cached-svg-image-with-css-cross-domain.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97633 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62739 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22208 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15770 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58323 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92160 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15825 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116717 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35441 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26108 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91331 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35814 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93908 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91132 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23228 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36020 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13790 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31188 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35344 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40882 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35057 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38410 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c6ca255d6e66933fbec2a5886c6d9731f55f6fc4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ea12985d19a77d34cc4da728c2dd8edbe7e8a309. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/03120f4eb67d38f4d7d27d2648cbf3c9a13357a5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/95d2e6196932a5dfab0dcfd14157d5e136e77118. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b40c0b83d4b3ad752320ff0198e863a5e2a501aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d964cd01b1c2ab81f6b1f1d7c2dee54ebfbff846. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9974033a603c7206123021c222219d0a405cd9a8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46816

smfr committed:
Make subpixel-positioned iframes paint and hit-test correctly
https://bugs.webkit.org/show_bug.cgi?id=294571
rdar://152114545

Reviewed by Alan Baradlay.

IFrames are Widgets, and Widgets live in an integer-based hierarchy, so have historically
painted and hit-tested using integer coordinates. However, iframes are often positioned on
subpixel boundaries, which leads to pixel gaps. Some of these are very obvious, e.g. an
iframe inside a `position:fixed; bottom: 0` element that should align seamlessly with UI.

However, we can adjust iframes for painting and hit-testing to overcome the integral
Widget limitation. This PR does this in various places.

First, `RenderWidget::paintContents()` adjusts the CTM to account for a subpixel position,
which fixes painting gaps (tested by iframe-subpixel-painting.html).

Second, the compositing code needs to handle subpixel positioning of the iframe's layers,
fixed in `RenderLayerBacking::updateAfterWidgetResize()` and tested by
iframe-subpixel-compositing.html.

Finally, hit-testing needs to maintain fractional coordinates;
`documentPointForWindowPoint()` is fixed to use FloatPoint coordinate mapping, and
HitTestLocation no longer rounds the origin of m_boundingBox which is what's used by the
`nodeAtPoint()` code.

Removing the flooring of the point in the HitTestLocation constructor caused
`css/css-transforms/transform-scale-hittest.html` to fail; this hit-tests a pixel above an
element with a 100x scale. The existing code would map the 1x1px hit-test area through the
scale, resulting in a 100x100px hit-test area; fix to map through the inverse transform to
maintain the 1x1px hit-test area in screen space. This is a merge of
https://chromium.googlesource.com/chromium/src.git/+/c47a6f0b4a592c2383fb553b6169f89ab01d851d

* LayoutTests/TestExpectations: imported/w3c/web-platform-tests/css/selectors/focus-visible-018-2.html
fails; https://github.com/web-platform-tests/wpt/pull/53241 exists to fix this.
* LayoutTests/compositing/iframes/hidpi-iframe-subpixel-compositing-expected.html: Added.
* LayoutTests/compositing/iframes/hidpi-iframe-subpixel-compositing.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-hit-test-expected.txt: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-hit-test.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-painting-expected.html: Added.
* LayoutTests/fast/frames/hidpi-iframe-subpixel-painting.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/elementFromPoint-subpixel-expected.txt:
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2-expected.txt:
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2.html: Tweak the test; browsers
have different results.
* Source/WebCore/page/EventHandler.cpp:
(WebCore::documentPointForWindowPoint):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/rendering/HitTestLocation.cpp:
(WebCore::HitTestLocation::HitTestLocation): No longer floor the point.
(WebCore::HitTestLocation::move): No longer convert to IntRect.
* Source/WebCore/rendering/HitTestingTransformState.cpp:
(WebCore::HitTestingTransformState::boundsOfMappedQuad const):
* Source/WebCore/rendering/HitTestingTransformState.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTestLayerByApplyingTransform):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterWidgetResize):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::frameViewDidChangeLocation):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::paintContents):

Canonical link: https://commits.webkit.org/296407@main
